library/core/src/iter/adapters/zip.rs

```diff
@@ -129,12 +185,83 @@ trait ZipImpl<A, B> {
     where
         A: DoubleEndedIterator + ExactSizeIterator,
         B: DoubleEndedIterator + ExactSizeIterator;
+    fn fold<T, F>(self, init: T, f: F) -> T
+    where
+        F: FnMut(T, Self::Item) -> T;
+    fn try_fold<T, F, R>(&mut self, init: T, f: F) -> R
+    where
+        F: FnMut(T, Self::Item) -> R,
+        R: Try<Output = T>;
+    fn rfold<T, F>(self, init: T, f: F) -> T
+    where
+        A: DoubleEndedIterator + ExactSizeIterator,
+        B: DoubleEndedIterator + ExactSizeIterator,
+        F: FnMut(T, Self::Item) -> T;
+    fn try_rfold<T, F, R>(&mut self, init: T, f: F) -> R
+    where
+        A: DoubleEndedIterator + ExactSizeIterator,
+        B: DoubleEndedIterator + ExactSizeIterator,
+        F: FnMut(T, Self::Item) -> R,
+        R: Try<Output = T>;
+
     // This has the same safety requirements as `Iterator::__iterator_get_unchecked`
     unsafe fn get_unchecked(&mut self, idx: usize) -> <Self as Iterator>::Item
     where
         Self: Iterator + TrustedRandomAccessNoCoerce;
 }
 
+#[inline]
+fn ok<B, T>(mut f: impl FnMut(B, T) -> B) -> impl FnMut(B, T) -> Result<B, !> {
+    move |acc, x| Ok(f(acc, x))
+}
+
+#[inline]
+fn check_rfold<AItem, B: DoubleEndedIterator, T, F: FnMut(T, (AItem, B::Item)) -> T>(
+    mut b: B,
+    mut f: F,
+) -> impl FnMut(T, AItem) -> T {
+    move |acc, x| match b.next_back() {
+        Some(y) => f(acc, (x, y)),
+        None => panic!("Iterator expected Some(item), found None."),
+    }
+}
+
+#[inline]
+fn check_try_fold<
```

Comment:
YMMV: this is a very long generic arguments list, so consider applying some APIT.  For example, maybe it could be

```rust
fn check_try_fold<
    'b,
    AItem,
    BItem,
    T,
    R: Try<Output = T>,
>(
    b: &'b mut impl Iterator<Item = BItem>,
    mut f: impl 'b + FnMut(T, (AItem, B::Item)) -> R,
) -> impl 'b + FnMut(T, AItem) -> ControlFlow<R, T> {
```

(Up to whether you think it's better, though.)